src/main/java/com/hubspot/jinjava/util/ScopeMap.java

```diff
@@ -120,7 +126,18 @@ public V remove(Object key) {
   }
 
   @Override
-  public void putAll(Map<? extends K, ? extends V> m) {
+  public void putAll(@Nonnull Map<? extends K, ? extends V> m) {
+    for (Entry<? extends K, ? extends V> entry : m.entrySet()) {
```

Comment:
Since you are iterating the entries, can you just call `put(key, value)` for each of the entry, and skip the `scope.putAll(m)`?

Reply:
The `HashMap` implementation of `putAll` has some optimizations for inserting multiple entries, so I didn't want to override that.